design/design-doc-resource-configurations.md

```diff
@@ -0,0 +1,838 @@
+spec:
+  forProvider:
+    network: /projects/foo/global/networks/bar
+    networkRef:
+      name: foobar
+    networkSelector:
+      matchConfiguration: true
+      matchLabels:
+        example: label
+```
+
+The above example shows a GCP `Subnetwork` managed resource that uses a
+reference selector. `networkSelector` sets the `networkRef` which in turn sets
+the value of `network`. The reference selector contains two optional fields:
+
+* `matchConfiguration`: Match only managed resources that are part of the same
+  `ResourceConfiguration` (i.e. have the same `configurationRef`) as the
+  selecting resource.
+* `matchLabels`: Match only managed resources with the supplied labels.
+
+If a reference field is set, its corresponding selector field is ignored. If the
+selector field is unset, it is ignored. If the specified selector matches
+multiple managed resources one is chosen at random. Note however that setting
+both `matchConfiguration` and `matchLabels` can guarantee that at most one
+provisioned managed resource will match the selector.
+
+### External Names
+
+Managed resources have a name and an _external name_. The former identifies the
+managed resource in the Kubernetes API, while the latter identifies the resource
+in the external system. Managed resources that have reached `v1beta1` allow both
+managed resource and resource claim authors to [control the name of their
+external resource] using the `crossplane.io/external-name` annotation:
+
+* The name of the external resource is set to the value of the annotation.
+* If the annotation is absent it is set to the managed resource's name.
+* The annotation is propagated from a resource claim to any managed resource it
+  dynamically provisions.
+
+This final point may pose a problem when one resource claim may provision many
+managed resources. If a resource claim set the `crossplane.io/external-name`
+annotation and referenced a `ResourceConfigurationClass` that specified two
+managed resources of the same kind, both dynamically provisioned resources would
+attempt to use the same external name. External names are unique at different
+scopes (global, project, region) for different resources so this will not be an
+issue all resources, but it may for some.
+
+This document proposes two new external naming annotations be introduced to help
+configuration class authors avoid external name conflicts:
+
+* `crossplane.io/external-name-prefix` - A value prepended to the external name
+  of a managed resource.
+* `crossplane.io/external-name-suffix` - A value appended to the external name
+  of a managed resource.
+
+These annotations would be set on the provisioned managed resources and parsed
+by their controllers alongside the existing `crossplane-io/external-name`
+annotation, resulting in an actual external name of the form
+`{external-name-prefix}{external-name}{external-name-suffix}`.
+
+### Connection Secrets
+
+TODO(negz): Determine a solution for connection secrets.
```

Comment:
Claims allow only one secret name to specify, it could make sense to keep that and prefix the keys. But prefix with what? Using managed resource kind would kill the portability. That's a hard one to get right.

Reply:
Yeah, this is probably the part of the design I have the least ideas about.

I'm increasingly wondering if we could get away with simply requiring the `ResourceConfiguration` to specify at most one managed resource whose connection secret will be propagated to the bound claim? If you buy my assertion that a resource claim represents an application's need for a thing (rather than being a generic abstraction for infrastructure), I wonder how likely it is that a claim will bind to a `ResourceConfiguration` consisting of multiple managed resources that publish connection secrets.

The two examples in this document will probably only have the one secret - GKE nodes don't have connection secrets, and Azure virtual network rules don't have connection secrets. One example of a likely `ResourceConfiguration` that had multiple secrets would be a configuration of an S3 bucket and an IAM user with access to said bucket.

Reply:
Perhaps we need to evolve the connection secret concept? Have each managed resource kind publish the connection keys they have available, then let a `ResourceConfiguration` surface those?

```yaml
---
kind: CoolManagedResource
metadata:
  name: cool
  labels:
    danksecrets: "true"
# ...
status:
  availableConnectionDetails:
  - username
  - password
  - endpoint
---
kind: ResourceConfiguration
# ...
spec:
  connectionDetailSelector:
    matchLabels:
      # Must match exactly one resource?
      danksecrets: "true"
    selectDetails:
    - username
    - password
    - endpoint
```

Reply:
> I'm increasingly wondering if we could get away with simply requiring the `ResourceConfiguration` to specify at most one managed resource whose connection secret will be propagated to the bound claim?

This still feels like a bold assumption.

> Have each managed resource kind publish the connection keys they have available, then let a ResourceConfiguration surface those?

I think since there is a high conflict possibility there has to be some pairing between the name of the keys that exist in managed resource secret and their names in connection secret that is exposed. But then the application loses the portability again, it would work only if the `ResourceConfigurationClass` that is used designed the keys in a way that app could work.

I think we can get away with prefixing the keys with claim kind but not necessarily the kind of the claim that is bound to `ResourceConfiguration`. Currently we have an implicit pairing between some managed resources and claims like `KubernetesCluster->GKECluster` and `MySQLInstance->CloudSQLInstance` so we could use that claim kind name for prefixing regardless of what claim the `ResourceConfiguration` is bound to.

It's the case that only claimable resources have connection secrets and this could be enforced; we can enforce that if your managed resource publishes a connection secret you have to pair it with a claim. Then the secret for a `KubernetesCluster` claim that have `GKECluster` and `CloudSQLInstance` could look like this:
```yaml
data:
  kubernetescluster.endpoint: 1.2.3.4
  ...
  mysqlinstance.endpoint: 3.2.4.1
  ...
```
We'd need to think about how we could implement this but it should be doable. The part where I don't feel super good about is to have to enforce that managed resources that publish a secret would have to have a claim type and correspondingly a `*Class` type, which sounds fine but introduces some dependency on the claim types that exist in the upstream. There is always the option to implement a claim for the infra-owner, but it's not very easy (`ResourceClaimDefinition` was solving this).

Reply:
> Currently we have an implicit pairing between some managed resources and claims

Would this still work when we're claiming a `ResourceConfiguration` using arbitrary, user defined claims?

Reply:
> But then the application loses the portability again, it would work only if the ResourceConfigurationClass that is used designed the keys in a way that app could work.

Do you feel like we really have portability at the connection secret level at the moment? For example I think we have a whole bunch of keys under the `CloudSQLInstance` secret that differ from (or don't exist in) the `RDSInstance` secret. I also suspect we're going to be increasingly moving outside of resources where portability is even possible - resources that aren't wire compatible across clouds.

Perhaps we need to allow `ResourceConfiguration` to define some mapping from the connection secret(s) of the managed resources to its own, so that you can declare the `hostname` field of the `Foo` managed resource connection secret to be the `endpoint` field of the `ResourceConfiguration` connection secret.

Reply:
> Would this still work when we're claiming a ResourceConfiguration using arbitrary, user defined claims?

Are there user-defined claims anymore in this proposal? Though this wouldn't work for the resources that publish a connection secret and don't have a claim, which is something we don't have right now but it's not safe to assume we won't.

> Do you feel like we really have portability at the connection secret level at the moment? For example I think we have a whole bunch of keys under the CloudSQLInstance secret that differ from (or don't exist in) the RDSInstance secret. I also suspect we're going to be increasingly moving outside of resources where portability is even possible - resources that aren't wire compatible across clouds.

As long as we don't have an intermediary application for every claim as an abstraction layer, we won't support 100% portability anyway. So, I think the best we can do is to aim for it optimistically and enable users to specify how portable they want it to be. For example, if they use only AKS and GKE, they might already have some bindings that are specific to those two resources in their applications and we'd allow them to configure the infra to meet what they already have.

Having a set of bindings looks like the simplest and the most explicit solution here. If they don't specify any bindings, all keys would go into the same secret, allowing override. If they do, they can change the key names in the target but it'd still be the same secret.

Reply:
I believe we now have a reasonable solution for connection secrets (thanks to @muvaf's work); we essentially establish a contract and mapping between composed kinds that can publish a connection secret and composite kinds that must surface those connection secrets.